### ISSUE TITLE
mode won't switch live

### ISSUE THREAD

**Issue by esponges** [srmklive/laravel-paypal#381]
Hey guys,

As many others, been working with this awesome package for some months, yet I've been struggling to work with live mode.

I've been all over the issues others had and elsewhere and haven't been able to find a solution for this.

If I remove sandbox credentials I get an infinite loop right after `dd($response)` from `$response = $provider->setExpressCheckout($checkoutData);
`

I'm using 1.0 so `ExpressCheckout` should not be the problem

I've tested on live server with same results.

My config

```
return [
    'mode'    => 'live', // Can only be 'sandbox' Or 'live'. If empty or invalid, 'live' will be used.
    'sandbox' => [
        'username'    => env('PAYPAL_SANDBOX_API_USERNAME', ''),
        'password'    => env('PAYPAL_SANDBOX_API_PASSWORD', ''),
        'secret'      => env('PAYPAL_SANDBOX_API_SECRET', ''),
        'certificate' => env('PAYPAL_SANDBOX_API_CERTIFICATE', ''),
        'app_id'      => 'APP-80W284485P519543T', // Used for testing Adaptive Payments API in sandbox mode
    ],
    'live' => [
        'username'    => env('PAYPAL_LIVE_API_USERNAME', ''),
        'password'    => env('PAYPAL_LIVE_API_PASSWORD', ''),
        'secret'      => env('PAYPAL_LIVE_API_SECRET', ''),
        'certificate' => env('PAYPAL_LIVE_API_CERTIFICATE', ''),
        'app_id'      => '', // Used for Adaptive Payments API
    ],

    'payment_action' => 'Sale', // Can only be 'Sale', 'Authorization' or 'Order'
    'currency'       => env('PAYPAL_CURRENCY', 'MXN'),
    'billing_type'   => 'MerchantInitiatedBilling',
    'notify_url'     => '', // Change this accordingly for your application.
    'locale'         => env('PAYPAL_LOCALE', 'es_ES'), // force gateway language  i.e. it_IT, es_ES, en_US ... (for express checkout only)
    'validate_ssl'   => false, // Validate SSL when creating api client.
];

```

env and env.example

```
#PayPal Setting & API Credentials - sandbox
PAYPAL_SANDBOX_API_USERNAME=sb-tr4z02598960_api1.business.example.com
PAYPAL_SANDBOX_API_PASSWORD=VSYUE32AU7MT7VY4
PAYPAL_SANDBOX_API_SECRET=ACdNh.ieqXXfGzIFBkj6F-fUdA49AIBmvpyOyz5MhLCsLyQVdcB74zVJ
PAYPAL_SANDBOX_API_CERTIFICATE=

#PayPal Setting & API Credentials - live
PAYPAL_LIVE_API_USERNAME=********************api1.gmail.com
PAYPAL_LIVE_API_PASSWORD=**********************K9W
PAYPAL_LIVE_API_SECRET=******************************FCBABE
PAYPAL_LIVE_API_CERTIFICATE= storage_path('cert_key_pem.txt'); // also tested empty
```

I use the new API credentials provided by PayPal

![image](https://user-images.githubusercontent.com/68973307/94625052-75798d80-027d-11eb-83a2-7c37864907cc.png)

Checked many other posts aswell as documentation and haven't been able to find solution

https://github.com/srmklive/laravel-paypal/issues/242
https://github.com/srmklive/laravel-paypal/issues/173
https://github.com/srmklive/laravel-paypal/issues/301

Thank you everyone else for your help.

**Comment by esponges** [comment]
I followed the flow and found that `setExpressCheckoutOptions($credentials)` method was not passing config and .env credentials so I passed manually in there:

```
public function setExpressCheckoutOptions($credentials)
    {
        $this->mode = 'live';
        // Setting API Endpoints
        if ($this->mode === 'sandbox') {
            $this->config['api_url'] = !empty($this->config['secret']) ?
                'https://api-3t.sandbox.paypal.com/nvp' : 'https://api.sandbox.paypal.com/nvp';

            $this->config['gateway_url'] = 'https://www.sandbox.paypal.com';
            $this->config['ipn_url'] = 'https://ipnpb.sandbox.paypal.com/cgi-bin/webscr';
        } else {
            $this->config['api_url'] = !empty($this->config['secret']) ?
                'https://api-3t.paypal.com/nvp' : 'https://api.paypal.com/nvp';

            $this->config['gateway_url'] = 'https://www.paypal.com';
            $this->config['ipn_url'] = 'https://ipnpb.paypal.com/cgi-bin/webscr';
        }

        // Adding params outside sandbox / live array
        $this->config['payment_action'] = $credentials['payment_action'];
        $this->config['notify_url'] = $credentials['notify_url'];
        /* Hardcode since not getting information from $credentials @ config.php and .env file */
        $this->config['locale'] = 'es_ES';
        $this->config['username'] = '************************gmail.com';
        $this->config['password'] = '*****************FK9W';
        $this->config['secret'] = '***********************************CBABE';
        $this->config['signature'] = '***********************************CBABE';

    }
```

After that I'd get this `dd($response)` from `setExpressCheckout`  method:

```
array:10 [▼
  "TIMESTAMP" => "2020-10-01T15:03:21Z"
  "CORRELATIONID" => "69bf453b7a011"
  "ACK" => "Failure"
  "VERSION" => "123"
  "BUILD" => "54906389"
  "L_ERRORCODE0" => "10002"
  "L_SHORTMESSAGE0" => "Security error"
  "L_LONGMESSAGE0" => "Security header is not valid"
  "L_SEVERITYCODE0" => "Error"
  "paypal_link" => null
]
```
Wich hasn't solved the issue.

Edit:

This has to do with credentials which takes me to what mentioned before.

_'¿Is **fingerprint** the same than **signature** which means it has to be set into the **SECRET** ??_ 

Using Secret from Developers Dashboard is not working

![image](https://user-images.githubusercontent.com/68973307/94833070-4d507280-03d4-11eb-9bfd-2415162783fb.png)

**Comment by srmklive** [comment]
signature is meant to be secret

**Comment by srmklive** [comment]
> I followed the flow and found that `setExpressCheckoutOptions($credentials)` method was not passing config and .env credentials so I passed manually in there:
> 
> ```
> public function setExpressCheckoutOptions($credentials)
>     {
>         $this->mode = 'live';
>         // Setting API Endpoints
>         if ($this->mode === 'sandbox') {
>             $this->config['api_url'] = !empty($this->config['secret']) ?
>                 'https://api-3t.sandbox.paypal.com/nvp' : 'https://api.sandbox.paypal.com/nvp';
> 
>             $this->config['gateway_url'] = 'https://www.sandbox.paypal.com';
>             $this->config['ipn_url'] = 'https://ipnpb.sandbox.paypal.com/cgi-bin/webscr';
>         } else {
>             $this->config['api_url'] = !empty($this->config['secret']) ?
>                 'https://api-3t.paypal.com/nvp' : 'https://api.paypal.com/nvp';
> 
>             $this->config['gateway_url'] = 'https://www.paypal.com';
>             $this->config['ipn_url'] = 'https://ipnpb.paypal.com/cgi-bin/webscr';
>         }
> 
>         // Adding params outside sandbox / live array
>         $this->config['payment_action'] = $credentials['payment_action'];
>         $this->config['notify_url'] = $credentials['notify_url'];
>         /* Hardcode since not getting information from $credentials @ config.php and .env file */
>         $this->config['locale'] = 'es_ES';
>         $this->config['username'] = '************************gmail.com';
>         $this->config['password'] = '*****************FK9W';
>         $this->config['secret'] = '***********************************CBABE';
>         $this->config['signature'] = '***********************************CBABE';
> 
>     }
> ```
> 
> After that I'd get this `dd($response)` from `setExpressCheckout` method:
> 
> ```
> array:10 [▼
>   "TIMESTAMP" => "2020-10-01T15:03:21Z"
>   "CORRELATIONID" => "69bf453b7a011"
>   "ACK" => "Failure"
>   "VERSION" => "123"
>   "BUILD" => "54906389"
>   "L_ERRORCODE0" => "10002"
>   "L_SHORTMESSAGE0" => "Security error"
>   "L_LONGMESSAGE0" => "Security header is not valid"
>   "L_SEVERITYCODE0" => "Error"
>   "paypal_link" => null
> ]
> ```
> 
> Wich hasn't solved the issue.
> 
> Edit:
> 
> This has to do with credentials which takes me to what mentioned before.
> 
> _'¿Is **fingerprint** the same than **signature** which means it has to be set into the **SECRET** ??_
> 
> Using Secret from Developers Dashboard is not working
> 
> ![image](https://user-images.githubusercontent.com/68973307/94833070-4d507280-03d4-11eb-9bfd-2415162783fb.png)

If you even bothered to read the documentation for v1, you would see you are totally wrong and quite frankly sound childish here.  

In order to override default api credentials, you need to use the `setApiCredentials` method, and provide it the configuration just as you described.

```
$provider->setApiCredentials($config);
```

In order to override the default parameters before an api call, just call the `addOptions` method to add your request data before the request to PayPal API:

```
$provider->addOptions($data_add)->setExpressCheckout($data);
```

**Comment by esponges** [comment]
Thanks for your kind response. Never meant to be disrespectfull with your work by -mistakenly- not reading or missunderstanding -I'm obviously a rookie in this- your documentation.

I managed to pass the credentials correctly. 

The credentials shown above were not correct.

My mistake was that when Requesting API Credentials I made a request for a API certificate _instead_ an API Signature. After selecting API signature I was given the right credentials.

![image](https://user-images.githubusercontent.com/68973307/95261966-f3471700-07f0-11eb-8048-3fec8889198d.png)

![image](https://user-images.githubusercontent.com/68973307/95262043-0eb22200-07f1-11eb-8187-259836567889.png)

Hopefully this might help someone in the future.
